summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/FutureQueue.scala

```diff
@@ -0,0 +1,202 @@
+   * Future completes.
+   *
+   * Returns the number of uncompleted Futures.  All results can
+   * still be dequeued, but only uncompleted Futures will count
+   * against the maxWaitingFutures limit.
+   */
+  def addAll(iter: TraversableOnce[(S, Future[T])]): Int = {
+    val addedSize = iter.count {
+      case (state, fut) => add(state, fut)
+    }
+
+    if (outstandingFutures.size > maxWaitingFutures.get) {
+      /*
+       * This can happen on large key expansion.
+       * May indicate maxWaitingFutures is too low.
+       */
+      logger.debug(
+        "Exceeded maxWaitingFutures({}), put {} futures", maxWaitingFutures.get, addedSize
+      )
+    }
+
+    addedSize
+  }
+
+  /**
+   * Queue a collection Future.  On failure, the state can be retrieved
+   * with the failure.  On success, the results are queued via addAll.
+   *
+   * The Future given here counts against the maxWaitingFutures, so calls
+   * to this method may cause a wait on dequeue.
+   */
+  def addAllFuture(state: S, iterFut: Future[TraversableOnce[(S, Future[T])]]): Unit =
+    addOutstandingFuture(
+      iterFut.respond {
+        case Return(iter) => addAll(iter)
+        case Throw(ex) => responses.put((state, Failure(ex)))
+      }.unit
+    )
+
+  private def addOutstandingFuture(fut: Future[Unit]): Boolean =
+    if (!fut.isDefined) {
+      numPendingOutstandingFutures.incrementAndGet
+      val ensured = fut.ensure(numPendingOutstandingFutures.decrementAndGet)
+      outstandingFutures.put(ensured)
+      true
+    } else {
+      false
+    }
+
+  private def forceExtraFutures() {
+    val maxWaitingFuturesCount = maxWaitingFutures.get
+    val pendingFuturesCount = numPendingOutstandingFutures.get
+    if (pendingFuturesCount > maxWaitingFuturesCount) {
+      // Too many futures waiting, let's clear.
+      val pending = outstandingFutures.toSeq.filterNot(_.isDefined)
+      val toClear = pending.size - maxWaitingFuturesCount
+      if (toClear > 0) {
+        try {
+          Await.ready(FutureQueue.waitN(pending, toClear), maxWaitingTime.get)
+        } catch {
+          case te: TimeoutException =>
+            logger.error(s"forceExtra failed on $toClear Futures", te)
+        }
+        outstandingFutures.putAll(pending.filterNot(_.isDefined))
+      } else {
+        outstandingFutures.putAll(pending)
+      }
+    } else {
+      // only dequeueAll if there's bang for the buck
+      if (outstandingFutures.size >= FutureQueue.OutstandingFuturesDequeueRatio * pendingFuturesCount) {
+        outstandingFutures.dequeueAll(_.isDefined)
+      }
+    }
+  }
+
+  /**
+   * Retrieve any completed results, along with their associated states.
+   *
+   * Returns up to maxEmitPerExec items.
+   */
+  def dequeue: TraversableOnce[(S, Try[T])] = {
+    // don't let too many futures build up
+    forceExtraFutures()
+    // Take all results that have been placed for writing to the network
+    responses.take(maxEmitPerExec.get)
```

Comment:
can we not take `maxEmitPerExec` as a class param and instead take it as a method arg here:

```scala
def dequeue(max: Int): Iterable[(S, Try[T])] = ...
```
This simplifies the state of the class and allows the caller to have better control (no need for a uniform max).